pycbc/events/stat.py

```diff
@@ -1810,6 +1810,142 @@ def coinc_lim_for_thresh(self, s, thresh, limifo,
         return loglr
 
 
+class ExpFitFgBgKDEStatistic(ExpFitFgBgNormStatistic):
+    """
+    The ExpFitFgBgNormStatistic with a mass and spin weighting factor
+    determined by KDE
+
+    This is the same as the ExpFitFgBgNormStatistic except the likelihood
+    is multiplied by the ratio of signal KDE to template KDE over template
+    params.
+    """
+
+    def __init__(self, sngl_ranking, files=None, ifos=None, **kwargs):
+        """
+        Create a statistic class instance
+
+        Parameters
+        ----------
+        sngl_ranking: str
+            The name of the ranking to use for the single-detector triggers.
+        files: list of strs, needed here
+            A list containing the filenames of hdf format files used to help
+            construct the coincident statistics. The files must have a 'stat'
+            attribute which is used to associate them with the appropriate
+            statistic class.
+        ifos: list of strs, not used here
+            The list of detector names
+        """
+        ExpFitFgBgNormStatistic.__init__(self, sngl_ranking, files=files,
+                                         ifos=ifos, **kwargs)
+        # the stat file attributes are hard-coded as 'signal-kde_file'
+        # and 'template-kde_file'
+        parsed_attrs = [f.split('-') for f in self.files.keys()]
+        self.kde_names = [at[0] for at in parsed_attrs if
+                       (len(at) == 2 and at[1] == 'kde_file')]
+        assert sorted(self.kde_names) == ['signal', 'template'], "two files" \
+                              "are required & one of them should have attr" \
+                              "'signal' & other should have attr 'template'"
+        self.kde_by_tid = {}
+        for kname in self.kde_names:
+            self.assign_kdes(kname)
+        # this will hold the template ids of the events for the
+        # statistic calculation
+        self.curr_tnum = None
+
+    def assign_kdes(self, kname):
+        """
+        Extract values from KDE files
+
+        Parameters
+        -----------
+        kname: str
+            Used to label the kde files.
+        """
+        with h5py.File(self.files[kname+'-kde_file'], 'r') as kde_file:
+            self.kde_by_tid[kname+'_kdevals'] = kde_file['data_kde'][:]
+
+    def single(self, trigs):
+        """
+        Calculate the necessary  single detector  information, that is,
+        getting tnum values from single detector triggers.
+
+        Parameters
+        ----------
+        trigs:dict of numpy.ndarrays, h5py group (or similar dict-like object)
+            Dictionary-like object holding single detector trigger information
+
+        Returns
+        -------
+        numpy.ndarray
+            The array of single detector values
+        """
+        try:
+            tnum = trigs.template_num # exists if accessed via coinc_findtrigs
```

Comment:
This doesn't look like it is safe, and I think may not work as intended in either the offline or online modes.

In the online modes the "single" method is called when triggers are added (I think) and the rank_stat_coinc called later with a larger set of triggers. Therefore in that case self.curr_tnum will not correspond to the "stats" sent to logsignalrate. 

In the offline mode, I think that the "single" method is called for all of the single detector triggers, and then logsignalrate is called on all potential coincs. Again, I don't think that self.curr_tnum here then aligns with the "stats" sent to logsignalrate below.

I may have misunderstood this, but I think the template number will need to be added to the dtype of "trigs" and then in here we call ExpFitFgBgNormStatistic.single to get other parameters and then add template_id into this.

Reply:
@spxiwh The intention is that this works (and only offline!) in a similar way to the BBH statistic with the curr_mchirp variable. In coinc_findtrigs the triggers and coincs should be processed strictly one template at a time, so we are thinking that it should be sufficient to assign tnum when the singles are read for a given template prior to finding coincs.